### PR TITLE
Feature:  An Implementation of VDR for resolving DIDs against an Indy ledger

### PR DESCRIPTION
**Pull request by pfeairheller**
This PR contains:

- A VDR implementation that relies on the go wrapper for libindy-vdr

To use this extension you will need to have [libindy-vdr.so](https://github.com/hyperledger/indy-vdr) installed.  

Signed-off-by: Phil Feairheller <pfeairheller@gmail.com>

### CODE REVIEW

File: component/vdr/indy/go.mod [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	github.com/flimzy/diff v0.1.7 // indirect
+	github.com/flimzy/testy v0.1.17 // indirect
+	github.com/go-kivik/couchdb v2.0.0+incompatible // indirect
+	github.com/go-kivik/kivik v2.0.0+incompatible // indirect
+	github.com/go-kivik/kiviktest v2.0.0+incompatible // indirect
+	github.com/go-sql-driver/mysql v1.5.0 // indirect
+	github.com/gopherjs/gopherjs v0.0.0-20200217142428-fce0ec30dd00 // indirect
+	github.com/hyperledger/aries-framework-go v0.1.5-0.20201119203645-6dedee7a40f7
+	github.com/hyperledger/indy-vdr/wrappers/golang v0.0.0-20201031155907-5f437d26ed71
+	github.com/mr-tron/base58 v1.2.0
```
Comment (troyronda) [inline component/vdr/indy/go.mod:16]: the framework currently brings in another base58 library: https://github.com/hyperledger/aries-framework-go/blob/master/go.mod#L12

Comment (troyronda) [inline component/vdr/indy/go.mod:16]: It looks like you are using two different base58 imports in this PR.

github.com/btcsuite/btcutil/base58 and github.com/mr-tron/base58

Comment (pfeairheller) [reply component/vdr/indy/go.mod:16]: I blame IntelliJ!  :-)

File: scripts/check_lint.sh [hunk 2]
```diff
@@ -15,7 +15,7 @@ if [ -z "$lint_path" ]; then
 fi
 
 DOCKER_CMD=${DOCKER_CMD:-docker}
-GOLANGCI_LINT_IMAGE="golangci/golangci-lint:v1.31.0"
+GOLANGCI_LINT_IMAGE="canislabs/golangci-lint:latest"
```
Comment (troyronda) [inline scripts/check_lint.sh:18]: Could we please revert this change?

File: scripts/check_unit.sh [hunk 3]
```diff
@@ -18,6 +18,12 @@ if [ -z "$unit_tests_path" ]; then
   unit_tests_path=./
 fi
 
+if [ -z "${EXCLUDE_TEST_PATH}" ]; then
```
Comment (troyronda) [inline scripts/check_unit.sh:21]: @soluchok

File: scripts/check_lint.sh (hunk 2 above)
Comment (pfeairheller) [reply scripts/check_lint.sh:18]: So that's one of the changes I was referring to when I mentioned that this was going to be complicated.  The linter won't work unless the indy_vdr lib is accessible so I had to create another docker image with that library.  

I welcome any suggestions for alternative!

Comment (troyronda) [reply scripts/check_lint.sh:18]: @pfeairheller One improvement would be to make GOLANGCI_LINT_IMAGE a script argument, with the default being "golangci/golangci-lint:v1.31.0"

`GOLANGCI_LINT_IMAGE=${GOLANGCI_LINT_IMAGE:-golangci/golangci-lint:v1.31.0}`

then where you call `make lint` in GH Actions:
`GOLANGCI_LINT_IMAGE=<alternate image> make lint`

File: component/vdr/indy/go.mod [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+module github.com/hyperledger/aries-framework-go-ext/component/vdr/indy
+
+go 1.15
+
+require (
+	github.com/btcsuite/btcutil v1.0.1
+	github.com/flimzy/diff v0.1.7 // indirect
+	github.com/flimzy/testy v0.1.17 // indirect
+	github.com/go-kivik/couchdb v2.0.0+incompatible // indirect
```
Comment (troyronda) [inline component/vdr/indy/go.mod:15]: I wonder where these imports are originating since couchDB and mysql was moved out of the framework.

File: scripts/check_lint.sh (hunk 2 above)
Comment (troyronda) [reply scripts/check_lint.sh:18]: If the indy_vdr lib is fetchable, it could also be possible to create a docker image as a CI step
FROM golangci/golangci-lint

Comment (pfeairheller) [reply scripts/check_lint.sh:18]: Unfortunately, that lib is not yet fetchable.  I like your first suggestion and will take a shot at it.  I'll also open an issue with indy-vdr to publish their lib and see if I can help get that moving.

File: component/vdr/indy/creator.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+	"time"
+
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/hyperledger/aries-framework-go/pkg/doc/did"
+	vdrapi "github.com/hyperledger/aries-framework-go/pkg/framework/aries/api/vdr"
+)
+
+// Build builds a new DID Doc.
+func (r *VDR) Build(pubKey *vdrapi.PubKey, opts ...vdrapi.DocOpts) (*did.Doc, error) {
```
Comment (llorllale) [inline component/vdr/indy/creator.go:19]: How does this DID get written to the Indy network?

File: component/vdr/indy/vdr.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+// WithRefresh option configuring an initial refresh of the Indy pool.
+func WithRefresh(refresh bool) Option {
+	return func(opts *VDR) {
+		opts.Refresh = refresh
+	}
+}
+
+// WithIndyClient configures the VDR with an existing Indy Client.
+func WithIndyClient(client Client) Option {
```
Comment (llorllale) [inline component/vdr/indy/vdr.go:88]: Resolution seems to depend entirely on this client. Will an implementation be provided soon?

Comment (pfeairheller) [reply component/vdr/indy/vdr.go:88]: Yes, that client is implemented in https://github.com/hyperledger/indy-vdr/tree/master/wrappers/golang

File: component/vdr/indy/creator.go (hunk 5 above)
Comment (pfeairheller) [reply component/vdr/indy/creator.go:19]: The assumption is that applications will write DIDs to the ledger outside of the VDR using the wrapper for lib-indyvdr directly.

File: component/vdr/indy/go.mod (hunk 4 above)
Comment (pfeairheller) [reply component/vdr/indy/go.mod:15]: Done.

File: scripts/check_lint.sh (hunk 2 above)
Comment (pfeairheller) [reply scripts/check_lint.sh]: Next build will have the GOLANGCI_LINT_IMAGE passed as an env variable so the Indy VDR can specify the modified docker image with the lib.

File: component/vdr/indy/go.mod [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+go 1.15
+
+require (
+	github.com/btcsuite/btcutil v1.0.1
+	github.com/hyperledger/aries-framework-go v0.1.5-0.20201119203645-6dedee7a40f7
+	github.com/hyperledger/indy-vdr/wrappers/golang v0.0.0-20201031155907-5f437d26ed71
+	github.com/stretchr/testify v1.6.1
+)
+
+replace (
```
Comment (troyronda) [inline component/vdr/indy/go.mod:18]: @pfeairheller these replaces have been removed in the latest framework.

**Review by troyronda — APPROVED** [review]
